ext/zpages/test/tracez_data_aggregator_benchmark.cc

```diff
@@ -0,0 +1,380 @@
+      query_thread.join();
+    }
+  }
+
+  /*
+   * Calling the aggregation work function, locking as needed.
+   */
+  void Aggregate()
+  {
+    std::lock_guard<std::mutex> lock(mtx);
+    aggregator->AggregateSpans();
+  }
+
+  /*
+   * Starts a background thread to query for aggregated data every
+   * query_interval nanoseconds. This should affect performance for the
+   * aggregator, since aggregation work cannot be done while aggegation
+   * data is being requested. This simulates a user visiting or refreshing
+   * the Tracez webpage many times, except this function does and holds
+   * nothing in memory.
+   */
+  void StartPeriodicQueryThread(nanoseconds query_interval = nanoseconds(1))
+  {
+    run.store(true, std::memory_order_release);
+    query_thread = std::thread([this, query_interval]() {
+      while (run.load(std::memory_order_acquire))
+      {
+        std::unique_lock<std::mutex> lock(mtx);
+        auto aggregations = aggregator->GetAggregatedTracezData();
+        // Continue if query interval passed
+        cont.wait_for(lock, query_interval);
+      }
+    });
+  }
+
+private:
+  std::unique_ptr<TracezDataAggregator> aggregator;
+  // Keep queries running while peer is in memory
+  std::thread query_thread;
+  // Ensure queries and aggregations don't happen simultaneously
+  std::mutex mtx;
+  std::atomic<bool> run;
+  std::condition_variable cont;
+  
+};
+
+////////////////////////  FIXTURE FOR SHARED SETUP CODE ///////////////////
+
+/*
+ * This is a regular fixture that initializes the shared needed components
+ * to create spans and do the aggregation work for benchmarking.
+ */
+class TracezAggregator : public benchmark::Fixture
+{
+protected:
+  void SetTracerAggregatorPeer()
+  {
+    std::shared_ptr<TracezSpanProcessor> processor(new TracezSpanProcessor());
+    tracer = std::shared_ptr<opentelemetry::trace::Tracer>(new Tracer(processor));
+    aggregator_peer = std::unique_ptr<TracezDataAggregatorPeer>(
+        new TracezDataAggregatorPeer(processor));
+  }
+
+  void SetUp(const ::benchmark::State& state)
+  {
+    SetTracerAggregatorPeer();
+  }
+
+  std::unique_ptr<TracezDataAggregatorPeer> aggregator_peer;
+  // Tracer for creating spans
+  std::shared_ptr<opentelemetry::trace::Tracer> tracer;
+};
+
+/*
+ * This extend the previous fixture, adding in the periodic query work in
+ * the background to simulate users requesting the data on the frontend
+ */
+class TracezAggregatorFetch : public TracezAggregator
+{
+protected:
+  void SetUp(const ::benchmark::State& state) override
+  {
+    SetTracerAggregatorPeer();
+    aggregator_peer->StartPeriodicQueryThread();
+  }
+};
+
+//////////////////////////// BENCHMARK DEFINITIONS /////////////////////////////////
+
+//////////////////////// USER NEVER VISITS WEBPAGE //////////////////////////////
+
+/*
+ * Aggregator handing many spans with the same name, who end instantly. This
+ * checks the scenario where there's only one Tracez name and minimal sorting
```

Comment:
Could you explain what you mean by sorting of latencies?

Reply:
i reclarified in the docstring, and it's supposed to be explaining the aggregator performance may be affected by needing to have computer whether a span is in error, running, and latency (including within bands)

Reply:
Why do you say that the aggregator performance might be affected by needing to compute the type of span? Running spans are received in a separate container altogether and distinguishing between error and latency span is done with a single if statement. Are you seeing a significant variance in performance when you test with the same number of running, completed and error spans? I am just curious and I am sure you understand these tests better so feel free to push this through.

Reply:
thanks for pointing that out! yes there's that minimal effect, and there's also memory considerations which i forgot to mention. all 11 buckets filled for each span name means there's more overhead compared to only 1 latency bucket

Reply:
Thanks for the clarification, I forgot the tests measured memory as well.